gnosis/safe/api/base_api.py

```diff
@@ -0,0 +1,46 @@
+from abc import ABC
+from typing import Dict, Optional
+from urllib.parse import urljoin
+
+import requests
+
+from gnosis.eth.ethereum_client import EthereumClient, EthereumNetwork
+
+
+class SafeAPIException(Exception):
+    pass
+
+
+class SafeBaseAPI(ABC):
+    URL_BY_NETWORK: Dict[EthereumNetwork, str] = {}
+
+    def __init__(
+        self, network: EthereumNetwork, ethereum_client: Optional[EthereumClient] = None
+    ):
+        self.network = network
+        self.ethereum_client = ethereum_client
+        self.base_url = self.URL_BY_NETWORK[network]
+
+    @classmethod
+    def from_ethereum_client(
+        cls, ethereum_client: EthereumClient
+    ) -> Optional["SafeBaseAPI"]:
+        ethereum_network = ethereum_client.get_network()
+        if ethereum_network in cls.URL_BY_NETWORK:
+            return cls(ethereum_network, ethereum_client=ethereum_client)
```

Comment:
I was thinking about moving this to the `__init__`. Do we consider `SafeBaseAPI` to be in a valid state if the result from `from_ethereum_client` is `None`?

Reply:
No, it's not, we should raise an error. Problem is that if we move that to `__init__` we need to make `EthereumNetwork` optional. WDYT?

Reply:
I personally like the idea of having these network parameters optional since they are not needed to encode transactions. And there are some use cases which only want to make use of transaction encoding.

Reply:
Added an exception if network is not found and the possibility of setting a custom base URL